src/Compilers/Core/Portable/Diagnostic/Diagnostic_SimpleDiagnostic.cs

```diff
@@ -130,11 +130,15 @@ public override bool Equals(object obj)
 
             public override int GetHashCode()
             {
-                return Hash.Combine(_descriptor,
-                        Hash.Combine(_messageArgs.GetHashCode(),
-                         Hash.Combine(_location.GetHashCode(),
-                          Hash.Combine(_severity.GetHashCode(), _warningLevel)
-                        )));
+                int hash = 0;
+
+                hash = Hash.Combine(_location.GetHashCode(), 0);
+                hash = Hash.Combine((int)_severity, hash);
```

Comment:
_location cannot be null. All the call sites that create SimpleDiagnostic use "location ?? Location.None". Probably we should have the same assignment expression in the constructors of SimpleDiagnostic to ensure _location is never set to null.